src/gen_bpf.c

```diff
@@ -1222,111 +1228,246 @@ static struct bpf_blk *_gen_bpf_syscall(struct bpf_state *state,
+ * @param s_head the head of the linked list to be returned to the caller
+ * @param s_tail the tail of the linked list to be returned to the caller
+ *
+ * Wrapper function for sorting syscalls
+ *
+ */
+static void _sys_sort(struct db_sys_list *syscalls,
+		      struct db_sys_list **s_head,
+		      struct db_sys_list **s_tail,
+		      uint32_t optimize)
+{
+	if (optimize != 2)
+		_sys_priority_sort(syscalls, s_head, s_tail);
+	else
+		/* sort by number for the binary tree */
+		_sys_num_sort(syscalls, s_head, s_tail);
+}
+
+/**
+ * Count the number of syscalls in the filter
+ * @param db the filter DB
+ * @param db_secondary the secondary DB
+ *
+ * Count the total number of syscalls that are in the seccomp fitler
+ */
+static unsigned int _get_syscall_cnt(const struct db_filter *db,
+				     const struct db_filter *db_secondary)
+{
+	struct db_sys_list *s_iter;
+	unsigned int syscall_cnt = 0;
+
+	db_list_foreach(s_iter, db->syscalls) {
+		syscall_cnt++;
+	}
+
+	if (db_secondary != NULL) {
+		db_list_foreach(s_iter, db_secondary->syscalls) {
+			syscall_cnt++;
+		}
+	}
+
+	return syscall_cnt;
+}
+
+/**
+ * Generate the BPF instruction blocks for a given filter/architecture
+ * @param state the BPF state
+ * @param db the filter DB
+ * @param db_secondary the secondary DB
+ *
+ * Generate the BPF instruction block for the given filter DB(s)/architecture(s)
+ * and return a pointer to the block on succes, NULL on failure.  The resulting
+ * block assumes that the architecture token has already been loaded into the
+ * BPF accumulator.
+ *
+ */
+static struct bpf_blk *_gen_bpf_arch(struct bpf_state *state,
+				     const struct db_filter *db,
+				     const struct db_filter *db_secondary,
+				     uint32_t optimize)
+{
+	int rc, level, i, j;
+	unsigned int blk_cnt = 0, syscall_cnt = 0, bintree_levels, empty_cnt = 0;
+	bool acc_reset;
+	struct bpf_instr instr;
+	struct db_sys_list *s_head = NULL, *s_tail = NULL, *s_iter;
+	struct bpf_blk *b_head = NULL, *b_tail = NULL, *b_iter, *b_new,
+		       *b_bintree;
+	uint64_t *bintree_hashes = NULL, nxt_hsh;
+	unsigned int *bintree_syscalls = NULL;
+
+	state->arch = db->arch;
+
+	/* sort the syscall list */
+	syscall_cnt = _get_syscall_cnt(db, db_secondary);
+	_sys_sort(db->syscalls, &s_head, &s_tail, optimize);
+	if (db_secondary != NULL)
+		_sys_sort(db_secondary->syscalls, &s_head, &s_tail, optimize);
+
+	bintree_levels = _get_bintree_levels(syscall_cnt, optimize);
+
+	if (bintree_levels > 0) {
+		empty_cnt = ((unsigned int)(SYSCALLS_PER_NODE * 2) <<
+			     (bintree_levels - 1)) - syscall_cnt;
+		bintree_hashes = zmalloc(sizeof(uint64_t) * bintree_levels);
+		if (bintree_hashes == NULL)
+			goto arch_failure;
+
+		bintree_syscalls = zmalloc(sizeof(unsigned int) *
+					   (bintree_levels));
+		if (bintree_syscalls == NULL)
+			goto arch_failure;
+
+		for (i = 0; i < bintree_levels; i++) {
+			bintree_syscalls[i] = INVALID_SYSCALL;
+			bintree_hashes[i]   = INVALID_HSH;
```

Comment:
I realize that `_gen_bpf_arch()` is already a bit unwieldy but I wonder if we can move some of the binary tree creation into another function, much like you did with the sort?